File: archive/c/c/longest-palindromic-substring.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

#define MAX_LENGTH 1000

int expandAroundCenter(const char* s, int left, int right) {
    while (left >= 0 && right < strlen(s) && s[left] == s[right]) {
        left--;
        right++;
    }
    return right - left - 1;
}

bool longestPalindromicSubstring(const char* s, char* result) {
    if (s == NULL || strlen(s) == 0) {
        strcpy(result, "");
        return false;
    }

    int start = 0, maxLength = 0;
    int len = strlen(s);

    for (int i = 0; i < len; i++) {
        int len1 = expandAroundCenter(s, i, i);
        int len2 = expandAroundCenter(s, i, i + 1);
        int len = (len1 > len2) ? len1 : len2;

        if (len > maxLength) {
            start = i - (len - 1) / 2;
            maxLength = len;
        }
    }

    if (maxLength > 1) {
        strncpy(result, s + start, maxLength);
        result[maxLength] = '\0';
        return true;
    }

    strcpy(result, "");
    return false;
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        printf("Usage: please provide a string that contains at least one palindrome\n");
        return 1;
    }

    const char* input = argv[1];
    char result[MAX_LENGTH];

    if (longestPalindromicSubstring(input, result) && strlen(result) > 0) {
        printf("%s\n", result);
    } else {
        printf("Usage: please provide a string that contains at least one palindrome\n");
    }

    return 0;
}

File: archive/r/ruby/longest-palindromic-substring.rb
def expand(s, l, r)
  while l >= 0 && r < s.length && s[l] == s[r]
    l -= 1
    r += 1
  end

  [l + 1, r - l - 1] # start, length
end

def longest_palindromic_substring(s)
  return if s.to_s.empty?

  best_start = 0
  best_len = 1

  s.length.times do |i|
    start1, len1 = expand(s, i, i)
    start2, len2 = expand(s, i, i + 1)

    start, len = (len1 > len2) ? [start1, len1] : [start2, len2]

    if len > best_len
      best_start, best_len = start, len
    end
  end

  s.slice(best_start, best_len)
end

input = ARGV.first
result = longest_palindromic_substring(input)

abort("Usage: please provide a string that contains at least one palindrome") if result.nil? || result.length <= 1

puts result
